lib/rake/application.rb

```diff
@@ -113,7 +113,10 @@ def top_level
         elsif options.show_prereqs
           display_prerequisites
         else
-          top_level_tasks.each { |task_name| invoke_task(task_name) }
+          Dir.chdir(@rakefile_dir || original_dir) do |dir|
+            top_level_tasks.each { |task_name| invoke_task(task_name) }
+            Dir.chdir(dir)
+          end
```

Comment:
This is based on [my suggestion in Bug #15661](https://bugs.ruby-lang.org/issues/15661#note-15).